File: 0977-squares-of-a-sorted-array/977.py
class Solution:
    def sortedSquares(sef, nums: List[int]) -> List[int]:
        l, r = 0, len(nums) -1
        result = [0] * len(nums)
        pos = len(nums) - 1
    
        while l <= r:
            left_sq = nums[l] ** 2
            right_sq = nums[r] ** 2
            if left_sq > right_sq:
                result[pos] = left_sq
                l += 1
            else:
                result[pos] = right_sq
                r -= 1
            pos -= 1
    
        return result

File: 0977-squares-of-a-sorted-array/977.rs
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut result = vec![0; nums.len()];
        let (mut l, mut r) = (0, nums.len() - 1);
        let mut pos = nums.len() - 1;

        for pos in (0..nums.len()).rev() {
            let left_sq = nums[l] * nums[l];
            let right_sq = nums[r] * nums[r];
            if left_sq > right_sq {
                result[pos] = left_sq;
                l += 1;
            }
            else {
                result[pos]= right_sq;
                r -= 1;
            }
        }
        result;
    }
}
